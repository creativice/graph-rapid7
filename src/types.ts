import { IntegrationInstanceConfig } from '@jupiterone/integration-sdk-core';

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * The provider API client region used to authenticate requests.
   */
  clientRegion: string;

  /**
   * The provider API client secret used to authenticate requests.
   */
  clientAccessToken: string;
}

export class StatusError extends Error {
  constructor(
    readonly options: {
      statusCode: number;
      statusText: string;
      message?: string;
    },
  ) {
    super(options.message);
    this.name = 'StatusError';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export type Rapid7User = {
  id: string;
  email: string;
  first_name: string;
  last_name: string;
  timezone: string;
  status: string;
  platform_admin: boolean;
  federated: boolean;
  last_login: string;
};
