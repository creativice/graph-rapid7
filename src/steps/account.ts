import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../types';

export const ACCOUNT_ENTITY_KEY = 'entity:account';
export const ACCOUNT_ENTITY_TYPE = 'rapid7_account';

export async function fetchAccountDetails({
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const accountEntity = createIntegrationEntity({
    entityData: {
      source: {
        id: 'acme-unique-account-id',
        name: 'Example Co. Acme Account',
      },
      assign: {
        _key: 'acme-unique-account-id',
        _type: ACCOUNT_ENTITY_TYPE,
        _class: 'Account',
        mfaEnabled: true,
      },
    },
  });

  await Promise.all([
    jobState.addEntity(accountEntity),
    jobState.setData(ACCOUNT_ENTITY_KEY, accountEntity),
  ]);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    types: [ACCOUNT_ENTITY_TYPE],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
