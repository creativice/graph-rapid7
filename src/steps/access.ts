import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  generateRelationshipType,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../types';
import { ACCOUNT_ENTITY_KEY, ACCOUNT_ENTITY_TYPE } from './account';

const USER_ENTITY_TYPE = 'rapid7_user';

export async function fetchUsers({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(ACCOUNT_ENTITY_KEY)) as Entity;

  await apiClient.iterateUsers(async (user) => {
    const userEntity = createIntegrationEntity({
      entityData: {
        source: user,
        assign: {
          _type: USER_ENTITY_TYPE,
          _class: 'User',
          name: `${user.first_name} ${user.last_name}`,
          username: user.email,
          email: user.email,
          firstName: user.first_name,
          lastName: user.last_name,
        },
      },
    });

    await Promise.all([
      jobState.addEntity(userEntity),
      jobState.addRelationship(
        createDirectRelationship({
          _class: 'HAS',
          from: accountEntity,
          to: userEntity,
        }),
      ),
    ]);
  });
}

export const accessSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-users',
    name: 'Fetch Users',
    types: [
      USER_ENTITY_TYPE,
      generateRelationshipType('HAS', ACCOUNT_ENTITY_TYPE, USER_ENTITY_TYPE),
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchUsers,
  },
];
