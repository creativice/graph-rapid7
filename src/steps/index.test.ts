import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../types';
import { fetchUsers } from './access';
import { fetchAccountDetails } from './account';

const DEFAULT_CLIENT_REGION = 'eu';
const DEFAULT_CLIENT_ACCESS_TOKEN = 'eu';

const integrationConfig: IntegrationConfig = {
  clientRegion: process.env.CLIENT_REGION || DEFAULT_CLIENT_REGION,
  clientAccessToken:
    process.env.CLIENT_ACCESS_TOKEN || DEFAULT_CLIENT_ACCESS_TOKEN,
};

test('should collect data', async () => {
  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchUsers(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Account'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: false,
      properties: {
        _type: { const: 'rapid7_account' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
    },
  });

  expect(
    context.jobState.collectedEntities.filter((e) => e._class.includes('User')),
  ).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: false,
      properties: {
        _type: { const: 'rapid7_user' },
        firstName: { type: 'string' },
        lastName: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
    },
  });
});
