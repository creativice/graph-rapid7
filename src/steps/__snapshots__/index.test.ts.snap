// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "acme-unique-account-id",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "acme-unique-account-id",
            "name": "Example Co. Acme Account",
          },
        },
      ],
      "_type": "rapid7_account",
      "createdOn": undefined,
      "displayName": "Example Co. Acme Account",
      "id": "acme-unique-account-id",
      "mfaEnabled": true,
      "name": "Example Co. Acme Account",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "cd29ae9e-c734-4746-8f8f-228e74f08928",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "email": "teemu@creativice.com",
            "federated": false,
            "first_name": "Teemu",
            "id": "cd29ae9e-c734-4746-8f8f-228e74f08928",
            "last_login": "2020-08-21T07:51:16Z",
            "last_name": "Hirvonen",
            "platform_admin": true,
            "status": "ACTIVE",
            "timezone": "UTC",
          },
        },
      ],
      "_type": "rapid7_user",
      "active": false,
      "createdOn": undefined,
      "displayName": "Teemu Hirvonen",
      "email": "teemu@creativice.com",
      "firstName": "Teemu",
      "id": "cd29ae9e-c734-4746-8f8f-228e74f08928",
      "lastName": "Hirvonen",
      "name": "Teemu Hirvonen",
      "status": "ACTIVE",
      "username": "teemu@creativice.com",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "acme-unique-account-id",
      "_key": "acme-unique-account-id|has|cd29ae9e-c734-4746-8f8f-228e74f08928",
      "_toEntityKey": "cd29ae9e-c734-4746-8f8f-228e74f08928",
      "_type": "rapid7_account_has_user",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "rapid7_account",
    "rapid7_user",
    "rapid7_account_has_user",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;
